quic: Add QUIC App Layer - v4

sharksforarms opened the pull request:
Previous PR: https://github.com/OISF/suricata/pull/5622

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3440

Describe changes:
- Parses gquic version Q043-Q046 and logs a CYU hash
- Added a mix of unit and suricata-verify pcap tests
- Fails gracefully with unknown QUIC versions
- Other versions can be added later w/ pcaps + testing
- Logs a cyu hash and string to eve.json
  - Example ```{"timestamp":"2020-09-02T15:58:21.139196+0000","flow_id":1610311033233340,"pcap_cnt":1,"event_type":"quic","src_ip":"10.0.2.15","src_port":51989,"dest_ip":"172.217.164.142","dest_port":443,"proto":"UDP","quic":[{"cyu":{"hash":"910a5e3a4d51593bd59a44611544f209","string":"46,PAD-SNI-VER-CCS-UAID-TCID-PDMD-SMHL-ICSL-NONP-MIDS-SCLS-CSCT-COPT-IRTT-CFCW-SFCW"}]}}```

References:
   - https://docs.google.com/document/d/1WJvyZflAO2pq77yOLbp9NsGjC1CHetAXV8I0fQe-B_U/edit
   - https://docs.google.com/document/d/1g5nIXAIkN_Y-7XJW5K45IblHd_L2f5LTaDUDwvZ5L6g/edit
   - https://www.slideshare.net/shigeki_ohtsu/quic-overview
   - https://tools.ietf.org/html/draft-ietf-quic-transport-17#section-19.8
   - https://github.com/salesforce/GQUIC_Protocol_Analyzer/blob/master/src/gquic-protocol.pac
   - cyu: https://engineering.salesforce.com/gquic-protocol-analysis-and-fingerprinting-in-zeek-a4178855d75f

suricata-verify-pr: 368

catenacyber commented:
Could you rebase the suricata-verify PR to get CI happier ?

catenacyber commented on rust/src/quic/logger.rs:
```diff
@@ -0,0 +1,42 @@
+use super::quic::QuicTransaction;
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+
+fn log_template(tx: &QuicTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
```
This should be documented 
Maybe in doc/userguide/output/eve/eve-json-format.rst ?

catenacyber commented on rust/src/quic/quic.rs:
```diff
@@ -0,0 +1,338 @@
+                }
+            },
+            Err(e) => {
+                SCLogDebug!("Quic Header Parsing Error: {:?}", e);
```
Should we set an app-layer event in these error cases ?
Grep for example `MQTTEvent::InvalidQosLevel`

sharksforarms replied:
In this case a header cannot be parsed and a transaction cannot be created... were you thinking of something else ? In the case of mqtt, the header can still be parsed

catenacyber replied:
> In this case a header cannot be parsed and a transaction cannot be created... were you thinking of something else ? 

I was thinking generally about error cases.
Parsing error, either for the header or something else, may be interesting...

catenacyber commented on src/app-layer-quic.c:
```diff
@@ -0,0 +1,57 @@
+#endif
+}
+
+#ifdef UNITTESTS
```
nits empty ifdef

catenacyber commented on src/detect-quic-cyu-string.c:
```diff
@@ -0,0 +1,197 @@
+
+    const uint8_t *data;
+    uint32_t data_len;
+    if (rs_quic_tx_get_cyu_string(cbdata->txv, cbdata->local_id, &data, &data_len) == 0) {
```
Mismatch between `rs_quic_tx_get_cyu_string` signature and `cbdata->local_id` type
I think it is best to have them both uint32_t/u32

catenacyber reviewed (commented):
Here are some quick comments about completeness to keep you going

jlucovsky reviewed (commented):
The QUIC s-v test fails for me:
```
===> quic-cyu: Sub test #1: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'event_type': 'quic', 'pcap_cnt': 1, 'quic.cyu[0].hash': 'a46560
d4548108cf99308319b3b85346', 'quic.cyu[0].string': '46,PAD-SNI-STK-VER-CCS-NONC-AEAD-UAID-SCID-TCID-PDMD-SMHL-ICSL-NONP-PUBS-MIDS-SCLS-KEXS-XLCT-CSCT-COPT-CCRT
-IRTT-CFCW-SFCW'}}
```

jlucovsky commented on rust/src/quic/quic.rs:
```diff
@@ -0,0 +1,338 @@
+}
+
+impl QuicTransaction {
+    fn new(header: QuicHeader, data: QuicData) -> QuicTransaction {
```
We've been converting the ctors to use `-> Self` ... also s/QuicTransaction/Self` on L45

victorjulien commented:
@sharksforarms are you planning on addressing these comments? Would like to pull this in before we reach suri 7

sharksforarms commented on rust/src/quic/cyu.rs:
```diff
@@ -0,0 +1,195 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
```
I rather keep this information in commits.

sharksforarms commented:
> @sharksforarms are you planning on addressing these comments? Would like to pull this in before we reach suri 7

Sorry, working on limited time at the moment. Addressed most comments + rebased here: https://github.com/OISF/suricata/pull/6307
